model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java

```diff
@@ -0,0 +1,261 @@
+                return new MapTaskWithValue<K, V>(value) {
+                    @Override
+                    public void execute(MapStorage<K, V> map) {
+                        map.putIfAbsent(key, getValue());
+                    }
+                };
+            }
+        },
+        REMOVE {
+            @Override
+            protected <K, V> MapTask<K, V> taskFor(K key, V value) {
+                return new MapTaskWithValue<K, V>(null) {
+                    @Override
+                    public void execute(MapStorage<K, V> map) {
+                        map.remove(key);
+                    }
+                };
+            }
+        },
+        REPLACE {
+            @Override
+            protected <K, V> MapTask<K, V> taskFor(K key, V value) {
+                return new MapTaskWithValue<K, V>(value) {
+                    @Override
+                    public void execute(MapStorage<K, V> map) {
+                        map.replace(key, getValue());
+                    }
+                };
+            }
+        },
+        ;
+
+        protected abstract <K, V> MapTask<K, V> taskFor(K key, V value);
+
+    }
+
+    private boolean active;
+    private boolean rollback;
+    private final Map<K, MapTask<K, V>> tasks = new LinkedHashMap<>();
+    private final MapStorage<K, V> map;
+
+    public MapKeycloakTransaction(MapStorage<K, V> map) {
+        this.map = map;
+    }
+
+    @Override
+    public void begin() {
+        active = true;
+    }
+
+    @Override
+    public void commit() {
+        if (rollback) {
+            throw new RuntimeException("Rollback only!");
+        }
+
+        for (MapTask<K, V> value : tasks.values()) {
+            value.execute(map);
+        }
+    }
+
+    @Override
+    public void rollback() {
+        tasks.clear();
+    }
+
+    @Override
+    public void setRollbackOnly() {
+        rollback = true;
+    }
+
+    @Override
+    public boolean getRollbackOnly() {
+        return rollback;
+    }
+
+    @Override
+    public boolean isActive() {
+        return active;
+    }
+
+    /**
+     * Adds a given task if not exists for the given key
+     * @return {@code true} if a task for the given key already exists
+     * @throws IllegalStateException
+     */
+    private void addTask(MapOperation op, K key, V value) throws IllegalStateException {
+        log.tracev("Adding operation {0} for {1}", op, key);
+
+        K taskKey = key;
+        tasks.merge(taskKey, op.taskFor(key, value), MapTaskCompose::new);
+    }
+
+    // This is for possibility to lookup for session by id, which was created in this transaction
+    public V get(K key, Function<K, V> defaultValueFunc) {
+        MapTask current = tasks.get(key);
+        if (current != null) {
+            if (current instanceof MapTaskWithValue) {
+                return ((MapTaskWithValue<K, V>) current).getValue();
+            }
+            return null;
```

Comment:
This function is used, for example, to find out whether a client exists. What if there is a client with desired id in the store, and within this session, it was updated with some `MapTask` without value. This function will return `null` and the code will think the client doesn't exist. Should there be `defaultValueFunc` call on this line?

Reply:
Actually, there is currently no `MapTask`, all instances are `MapTaskWithValue`s, so this will never happen. Perhaps the code is worth simplifying.

Reply:
I agree, currently, it should not happen, but in future, somebody can add such a task. Anyway, it can stay as it is, I don't have any strong reason to change it, I just wanted to point it out.